packages/flutter/lib/src/material/scaffold.dart

```diff
@@ -61,6 +61,139 @@ enum _ScaffoldSlot {
   statusBar,
 }
 
+/// Doc
+class ScaffoldMessenger extends StatefulWidget {
+  /// Doc
+  ///
+  /// The [child] argument must not be null.
+  const ScaffoldMessenger({
+    Key key,
+    @required this.child,
+  }) : assert(child != null),
+      super(key: key);
+
+  /// Returns the closest [ScaffoldMessengerState] which encloses the given
+  /// context.
+  ///
+  /// Typical usage is as follows:
+  ///
+  /// ```dart
+  /// ScaffoldMessages scaffoldMessenger = ScaffoldMessages.of(context);
+  /// scaffoldMessenger.showSnackBar(const SnackBar(content: Text('Snack-tastic'));
+  /// ```
+  static ScaffoldMessengerState of(BuildContext context) {
+    final _ScaffoldMessengerScope scope = context.dependOnInheritedWidgetOfExactType<_ScaffoldMessengerScope>();
+    return scope?._scaffoldMessengerState;
+  }
+
+  /// The widget below this widget in the tree.
+  ///
+  /// {@macro flutter.widgets.child}
+  final Widget child;
+
+  @override
+  ScaffoldMessengerState createState() => ScaffoldMessengerState();
+}
+
+/// Doc
+class ScaffoldMessengerState extends State<ScaffoldMessenger> {
+  final Set<ScaffoldState> _scaffolds = <ScaffoldState>{};
+  final Queue<ScaffoldFeatureController<SnackBar, SnackBarClosedReason>> _snackBars = Queue<ScaffoldFeatureController<SnackBar, SnackBarClosedReason>>();
+
+  void _register(ScaffoldState scaffold) {
+    // Are we in the middle of showing a SnackBar elsewhere?
+    // If so, present on the new scaffold.
+    if (!_scaffolds.contains(scaffold) && _snackBars.isNotEmpty) {
+      scaffold.showSnackBar(_snackBars.first._widget, listener: _handleSnackBarStatusChange);
+    }
+    _scaffolds.add(scaffold);
+  }
+
+  void _unregister(ScaffoldState scaffold) {
+    _scaffolds.remove(scaffold);
```

Comment:
Presumably, this also will have to remove things from `_snackBars` if a snackbar was shown when the unregister happened?